presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreHttpRequestDetails.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.metastore.thrift;
+
+import com.facebook.airlift.json.ObjectMapperProvider;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class ThriftMetastoreHttpRequestDetails
+{
+    private final String url;
+    private final Map<String, String> headers;
+
+    @JsonCreator
+    public ThriftMetastoreHttpRequestDetails(@JsonProperty("url") String url,
+                                             @JsonProperty("headers") Map<String, String> headers)
+    {
+        this.url = url;
+        this.headers = headers;
+    }
+
+    @JsonCreator
+    public static ThriftMetastoreHttpRequestDetails fromString(@JsonProperty("hive.metastore.discovery-uri") String discoveryUri) throws TException
+    {
+        try {
+            return new ObjectMapperProvider().get().readValue(discoveryUri, ThriftMetastoreHttpRequestDetails.class);
```

Comment:
this is anti-pattern by using a json blob in the config; why don't we use something like `?user=XXX&endpoint=XXX` in the url?

Reply:
So this json structure was created because internally, at Uber, our http discovery uri expects the URL and 2 headers, RPC-Service and RPC-Caller. I can update the code to take these as query param inputs and later interpret it as headers, but how do we make it flexible enough for orgs to specify their own http url, which could have query param or headers or both?